File: VB/SubDocumentHelper.vb
Imports DevExpress.XtraRichEdit.API.Native
Imports System.Runtime.CompilerServices

Namespace SubDocumentIterator

    Public Module SubDocumentHelper

        Public Delegate Sub SubDocumentDelegate(ByVal subDocument As SubDocument)

        <Extension()>
        Public Sub ForEachSubDocument(ByVal document As Document, ByVal subDocumentProcessor As SubDocumentDelegate)
            subDocumentProcessor(document)
            ProcessShapes(document.Shapes, subDocumentProcessor)
            ProcessComments(document.Comments, subDocumentProcessor)
            For Each section As Section In document.Sections
                ProcessSection(section, HeaderFooterType.First, subDocumentProcessor)
                ProcessSection(section, HeaderFooterType.Odd, subDocumentProcessor)
                ProcessSection(section, HeaderFooterType.Even, subDocumentProcessor)
            Next
        End Sub

        Private Sub ProcessSection(ByVal section As Section, ByVal headerFooterType As HeaderFooterType, ByVal subDocumentProcessor As SubDocumentDelegate)
            If section.HasHeader(headerFooterType) Then
                Dim header As SubDocument = section.BeginUpdateHeader(headerFooterType)
                subDocumentProcessor(header)
                ProcessShapes(header.Shapes, subDocumentProcessor)
                section.EndUpdateHeader(header)
            End If

            If section.HasFooter(headerFooterType) Then
                Dim footer As SubDocument = section.BeginUpdateFooter(headerFooterType)
                subDocumentProcessor(footer)
                ProcessShapes(footer.Shapes, subDocumentProcessor)
                section.EndUpdateFooter(footer)
            End If
        End Sub

        Private Sub ProcessShapes(ByVal shapes As ShapeCollection, ByVal subDocumentProcessor As SubDocumentDelegate)
            For Each shape As Shape In shapes
                If shape.ShapeFormat.TextBox IsNot Nothing Then subDocumentProcessor(shape.ShapeFormat.TextBox.Document)
            Next
        End Sub

        Private Sub ProcessComments(ByVal comments As CommentCollection, ByVal subDocumentProcessor As SubDocumentDelegate)
            For Each comment As Comment In comments
                Dim commentSubDocument As SubDocument = comment.BeginUpdate()
                subDocumentProcessor(commentSubDocument)
                comment.EndUpdate(commentSubDocument)
            Next
        End Sub
    End Module
End Namespace

File: CS/SubDocumentHelper.cs
using DevExpress.XtraRichEdit.API.Native;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SubDocumentIterator
{
    public static class SubDocumentHelper
    {
        public delegate void SubDocumentDelegate(SubDocument subDocument);
        public static void ForEachSubDocument(this Document document, SubDocumentDelegate subDocumentProcessor)
        {
            subDocumentProcessor(document);
            ProcessShapes(document.Shapes, subDocumentProcessor);
            ProcessComments(document.Comments, subDocumentProcessor);
            foreach (Section section in document.Sections)
            {
                ProcessSection(section, HeaderFooterType.First, subDocumentProcessor);
                ProcessSection(section, HeaderFooterType.Odd, subDocumentProcessor);
                ProcessSection(section, HeaderFooterType.Even, subDocumentProcessor);
            }
        }
        private static void ProcessSection(Section section, HeaderFooterType headerFooterType, SubDocumentDelegate subDocumentProcessor)
        {
            if (section.HasHeader(headerFooterType))
            {
                SubDocument header = section.BeginUpdateHeader(headerFooterType);
                subDocumentProcessor(header);
                ProcessShapes(header.Shapes, subDocumentProcessor);
                section.EndUpdateHeader(header);
            }
            if (section.HasFooter(headerFooterType))
            {
                SubDocument footer = section.BeginUpdateFooter(headerFooterType);
                subDocumentProcessor(footer);
                ProcessShapes(footer.Shapes, subDocumentProcessor);
                section.EndUpdateFooter(footer);
            }
        }
        private static void ProcessShapes(ShapeCollection shapes, SubDocumentDelegate subDocumentProcessor)
        {
            foreach (Shape shape in shapes)
                if (shape.ShapeFormat.TextBox != null)
                    subDocumentProcessor(shape.ShapeFormat.TextBox.Document);
        }
        private static void ProcessComments(CommentCollection comments, SubDocumentDelegate subDocumentProcessor)
        {
            foreach (Comment comment in comments)
            {
                SubDocument commentSubDocument = comment.BeginUpdate();
                subDocumentProcessor(commentSubDocument);
                comment.EndUpdate(commentSubDocument);
            }
        }
    }
}
